sharedbuffers/mapped_struct.py

```diff
@@ -333,14 +335,185 @@ class mapped_dict(dict):
 
     @classmethod
     def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0):
-        keys = list(obj)
-        values = [obj[k] for k in keys]
-        return mapped_list.pack_into([keys, values], buf, offs, idmap, implicit_offs)
+        return proxied_dict.pack_into(obj, buf, offs, idmap, implicit_offs)
 
     @classmethod
     def unpack_from(cls, buf, offs, idmap = None):
-        key, values = mapped_list.unpack_from(buf, offs, idmap)
-        return cls(zip(key, values))
+        proxy = proxied_dict.unpack_from(buf, offs, idmap)
+        return {k: v for k, v in proxy.iteritems()}
+
+
+def _upsize(x):
+    "Round up X to the next power of 2"
+    for p2 in (1, 2, 4, 8, 16, 32, 64):
+        x |= x >> p2
+    return x + 1
+
+
+def _hash_rotl(code, nbits):
+    return (code << nbits) | (code >> (32 - nbits))
+
+
+def _mix_hash(code1, code2):
+    return _hash_rotl(code1, 5) ^ code2
+
+
+def _stable_hash(key):
+    if isinstance(key, basestring):
+        hval = xxhash.xxh64(key).intdigest()
+    elif isinstance(key, int):
+        hval = key
+    elif isinstance(key, float):
+        mant, expo = math.frexp(key)
+        hval = _mix_hash(expo, int(mant * 0xffffffffffff))
+    elif isinstance(key, (tuple, proxied_tuple, frozenset)):
+        hval = xxhash.xxh64(str(type(key))).intdigest()
```

Comment:
This is most certainly not a good idea. `str(type(key))` could be unstable.

Check `proxied_tuple`, you have an algorithm there I believe you can use. `proxied_tuple` uses python's hash as a base, but you can use the same algorithm replacing `hash` with `stable_hash`.